core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/tenant/JDBCTenantManager.java

```diff
@@ -1201,4 +1223,74 @@ private boolean checkUniqueIdColumnInTable() throws UserStoreException {
         }
         return false;
     }
+
+    /**
+     * Delete tenant UM Data.
+     *
+     * @param tenantId Id of the tenant
+     * @param conn DB connection
+     * @throws Exception
+     */
+    private static void deleteTenantUMData(int tenantId, Connection conn) throws Exception {
+
+        try {
+            conn.setAutoCommit(false);
+
+            executeDeleteQuery(conn, DBConstants.DELETE_USER_PERMISSIONS_BY_TENANT_ID_SQL, tenantId);
+            executeDeleteQuery(conn, DBConstants.DELETE_ROLE_PERMISSIONS_BY_TENANT_ID_SQL, tenantId);
+            executeDeleteQuery(conn, DBConstants.DELETE_PERMISSION_BY_TENANT_ID_SQL, tenantId);
+
+            executeDeleteQuery(conn, ProfileDBConstant.DELETE_CLAIM_BEHAVIOR_BY_TENANT_ID, tenantId);
+            executeDeleteQuery(conn, ProfileDBConstant.DELETE_PROFILE_CONFIG_BY_TENANT_ID_SQL, tenantId);
+
+            executeDeleteQuery(conn, ClaimDBConstants.DELETE_CLAIM_BY_TENANT_ID_SQL, tenantId);
+            executeDeleteQuery(conn, ClaimDBConstants.DELETE_DIALECT_BY_TENANT_ID_SQL, tenantId);
+
+            executeDeleteQuery(conn, JDBCRealmConstants.DELETE_USER_PROPERTY_BY_TENANT_ID_SQL, tenantId);
+            executeDeleteQuery(conn, JDBCRealmConstants.DELETE_DOMAIN_FROM_USER_ROLE_BY_TENANT_ID, tenantId);
+            executeDeleteQuery(conn, JDBCRealmConstants.REMOVE_USER_ROLES_BY_TENANT_ID_SQL, tenantId);
+            executeDeleteQuery(conn, JDBCRealmConstants.DELETE_ROLES_BY_TENANT_ID_SQL, tenantId);
+            executeDeleteQuery(conn, JDBCRealmConstants.DELETE_USERS_BY_TENANT_ID_SQL, tenantId);
+
+            executeDeleteQuery(conn, HybridJDBCConstants.DELETE_ROLES_BY_TENANT_ID_SQL, tenantId);
+            executeDeleteQuery(conn, HybridJDBCConstants.DELETE_REMEMBERME_VALUES_BY_TENANT_ID_SQL, tenantId);
+
+            executeDeleteQuery(conn, TenantConstants.DELETE_TENANT_SQL, tenantId);
+
+            conn.commit();
+        } catch (Exception e) {
+            conn.rollback();
+            String errorMsg = "An error occurred while deleting um data for tenant: " + tenantId;
+            throw new Exception(errorMsg, e);
+        } finally {
+            conn.close();
+        }
+    }
+
+    /**
+     * Execute deletion queries.
+     *
+     * @param conn DB connection
+     * @param query SQL query
+     * @param tenantId Id of the tenant
+     * @throws Exception
+     */
+    private static void executeDeleteQuery(Connection conn, String query, int tenantId)
```

Comment:
Shall we throw a specific exception from here? WDYT?

Reply:
Yes. It's better. I threw a new Exception here as I already handled the SQLException inside this function. Or we can throw the same SQLException from here too.